src/interp/interp-inl.h

```diff
@@ -190,6 +190,44 @@ auto FreeList<T>::count() const -> Index {
   return list_.size() - free_.size();
 }
 
+//// FreeRefList ////
+inline FreeRefList::Index FreeRefList::New(Ref ref) {
+  if (free_head_ == 0) {
+    list_.push_back(ref);
+    return list_.size() - 1;
+  }
+
+  Index index = free_head_ - 1;
+  assert(!IsUsed(index));
+  free_head_ = list_[index].index - unusedStart;
+  list_[index] = ref;
+  return index;
+}
+
+inline void FreeRefList::Delete(Index index) {
+  assert(IsUsed(index));
+  list_[index].index = free_head_ + unusedStart;
+  free_head_ = index + 1;
+}
+
+inline bool FreeRefList::IsUsed(Index index) const {
+  return list_[index].index < unusedStart;
+}
+
+inline const Ref& FreeRefList::Get(Index index) const {
+  assert(IsUsed(index));
+  return list_[index];
+}
+
+inline Ref& FreeRefList::Get(Index index) {
+  assert(IsUsed(index));
+  return list_[index];
+}
+
+inline FreeRefList::Index FreeRefList::size() const {
+  return list_.size();
+}
```

Comment:
Some of these methods look identical to `FreeList<T>`..  can we avoid the duplication somehow?    Partial template specialization maybe?

Reply:
I am a c/assembly dev and c++ magic is a bit confusing for me. FreeList<T> has fields which are not used for Refs and vice versa, and I don't know how can I make unused fields disappear. If FreeList<T> will be also reworked, specializing will be easier. Shall I do the whole rework in one step?